File: 2018/day17/solve.py
#!/usr/bin/env python3
from collections import defaultdict, namedtuple
from pathlib import Path
import re
import sys

Point = namedtuple("Point", ["x", "y"])
Rectangle = namedtuple("Rectangle", ["ul", "lr"])
Scan = namedtuple("Scan", ["view", "actual", "spring", "clays", "waters", "damps"])

UP = (0, -1)
DOWN = (0, 1)
LEFT = (-1, 0)
RIGHT = (1, 0)


class Solve:
    def __init__(self, verbose):
        self.verbose = verbose
        self.view = None
        self.actual = None
        self.spring = None
        self.data = defaultdict(lambda: ".")
        self.cache = set()

    @staticmethod
    def add_dir(t, dir):
        return (t[0] + dir[0], t[1] + dir[1])

    @staticmethod
    def intersects(rect, point):
        return (rect.ul.x <= point.x and rect.ul.y <= point.y and
                rect.lr.x >= point.x and rect.lr.y >= point.y)

    @staticmethod
    def parse_line(line):
        def parse_val(txt):
            m = re.match(r'([0-9]+)\.\.([0-9]+)', txt)
            if m:
                return (int(m.group(1)), int(m.group(2)))
            else:
                return (int(txt), int(txt))

        m = re.match(r'([xy])=([0-9.]+), ([xy])=([0-9.]+)', line)
        if not m:
            raise Exception("Bad line parse: " + line)
        return {m.group(1): parse_val(m.group(2)), m.group(3): parse_val(m.group(4))}

    @staticmethod
    def load_lines(file):
        fpath = Path(file)
        if not fpath.exists():
            return []
        with open(fpath) as f:
            return [line.strip() for line in f.readlines()]

    def parse_input(self, file):
        parsed = [self.parse_line(line) for line in self.load_lines(file)]
        min_x = min(p["x"][0] for p in parsed)
        max_x = max(p["x"][1] for p in parsed)
        min_y = min(p["y"][0] for p in parsed)
        max_y = max(p["y"][1] for p in parsed)

        self.view = Rectangle(ul=Point(x=min_x, y=min_y), lr=Point(x=max_x, y=max_y))
        # Expand out one square in each dir to handle overflows above the range
        # that fall back in and that kind of thing
        self.actual = Rectangle(
            ul=Point(x=min_x - 1, y=min_y - 1), lr=Point(x=max_x + 1, y=max_y))
        self.spring = Point(x=500, y=min_y-1)

        self.data[(self.spring.x, self.spring.y)] = "+"
        for p in parsed:
            for x in range(p["x"][0], p["x"][1] + 1):
                for y in range(p["y"][0], p["y"][1] + 1):
                    self.data[(x, y)] = "#"

    def print_scan(self):
        for y in range(self.actual.ul.y, self.actual.lr.y + 1):
            line = "".join(self.data.get((x, y), ".") for x in range(
                self.actual.ul.x, self.actual.lr.x + 1))
            print(line)
        print()

    def is_blocker(self, cur):
        return self.data.get(cur, ".") in ("#", "~")

    def is_empty(self, cur):
        return self.data.get(cur, ".") in (".", "|")

    def is_void(self, cur):
        return cur[1] > self.actual.lr.y

    # go dir until you encounter a wall or a gap. if a wall, stop; if a gap,
    # try to fill it, and continue if you did
    def send_sideways(self, drop, dir):
        while not self.is_blocker(drop):
            gap = self.add_dir(drop, DOWN)
            if self.is_empty(gap):
                self.send_drop(gap)
                if self.is_empty(gap):  # couldn't fill it
                    return (gap, None)
            drop = self.add_dir(drop, dir)
        return (None, drop)

    def send_drop(self, drop):
        if drop in self.cache:
            if self.verbose > 0:
                print(f"Sending drop from {drop} - cached, will return")
            return

        if self.verbose > 0:
            print(f"Sending drop from {drop}")

        t = self.data.get(drop, ".")
        # see if this drop already got covered in water by an earlier drop:
        if t not in (".", "|", "+"):
            if self.verbose > 1:
                print(f"Drop seems to be already underwater at {drop}")
            return

        floor = drop
        while not self.is_blocker(floor) and not self.is_void(floor):
            if floor != (self.spring.x, self.spring.y):
                self.data[floor] = "|"
            floor = self.add_dir(floor, DOWN)
        if self.is_void(floor):
            if self.verbose > 1:
                print(f"Drop fell off-screen at {floor}")
            self.cache.add(drop)
            return

        level = self.add_dir(floor, UP)
        while level[1] >= drop[1]:
            left_gap, left_wall = self.send_sideways(level, LEFT)
            right_gap, right_wall = self.send_sideways(level, RIGHT)
            if left_wall and right_wall:
                if self.verbose > 1:
                    print(f"Drop fills between {left_wall} and {right_wall}")
                cur = self.add_dir(left_wall, RIGHT)
                while cur != right_wall:
                    self.data[cur] = "~"
                    cur = self.add_dir(cur, RIGHT)
                # and continue to loop to fill more
                level = self.add_dir(level, UP)
            else:
                # can't fill any more
                if self.verbose > 1:
                    print(f"Drop falls through {left_gap or '<none>'} and/or {right_gap or '<none>'}")

                cur = level
                stop = self.add_dir(left_gap, UP) if left_gap else self.add_dir(left_wall, RIGHT)
                while cur != stop:
                    self.data[cur] = "|"
                    cur = self.add_dir(cur, LEFT)
                self.data[stop] = "|"

                cur = level
                stop = self.add_dir(right_gap, UP) if right_gap else self.add_dir(right_wall, LEFT)
                while cur != stop:
                    self.data[cur] = "|"
                    cur = self.add_dir(cur, RIGHT)
                self.data[stop] = "|"

                break
        self.cache.add(drop)

    def count_wet(self):
        return sum(1 for v in self.data.values() if v in ("~", "|"))

    def count_water(self):
        return sum(1 for v in self.data.values() if v in ("~"))

    def main(self, file):
        self.parse_input(file)
        self.print_scan()
        self.send_drop((self.spring.x, self.spring.y))
        self.print_scan()
        wet_num = self.count_wet()
        print(f"Wet squares: {wet_num}")
        water_num = self.count_water()
        print(f"Water squares: {water_num}")


if __name__ == "__main__":
    verbose = int(sys.argv[2]) if len(sys.argv) > 2 else 0
    Solve(verbose).main(sys.argv[1])

File: 2018/day17/solve.exs
defmodule Solve do
  defmodule Scan do
    defstruct data: %{}, view: {0, 0, 0, 0}, actual: {0, 0, 0, 0}, spring: {0, 0}
  end

  def verbose() do
    0
  end

  def parse_val(str) do
    case Regex.named_captures(~r/(?<from>[0-9]+)\.\.(?<to>[0-9]+)/, str) do
      %{"from" => from, "to" => to} -> {String.to_integer(from), String.to_integer(to)}
      _ -> {String.to_integer(str), String.to_integer(str)}
    end
  end

  def parse_line(line) do
    case Regex.named_captures(~r/(?<vn1>[xy])=(?<vv1>[0-9.]+), (?<vn2>[xy])=(?<vv2>[0-9.]+)/, line) do
      %{"vn1" => vn1, "vv1" => vv1, "vn2" => vn2, "vv2" => vv2} ->
        [%{vn1 => parse_val(vv1), vn2 => parse_val(vv2)}]
      _ ->
        if line != "" do
          IO.puts("Bad line parse: " <> line)
        end
        []
    end
  end

  def parse_input(file) do
    {:ok, data} = File.read(file)
    clays = data |> String.split("\n") |> Enum.flat_map(&parse_line/1)

    min_x = clays |> Enum.map(fn %{"x" => {m, _}, "y" => _} -> m end) |> Enum.min
    max_x = clays |> Enum.map(fn %{"x" => {_, m}, "y" => _} -> m end) |> Enum.max
    min_y = clays |> Enum.map(fn %{"y" => {m, _}, "x" => _} -> m end) |> Enum.min
    max_y = clays |> Enum.map(fn %{"y" => {_, m}, "x" => _} -> m end) |> Enum.max

    # Expand out one square in each dir to handle overflows above the range
    # that fall back in and that kind of thing
    base_data = for x <- (min_x - 1..max_x + 1), y <- (min_y - 1..max_y) do
      {{x, y}, "."}
    end

    clay_data = clays |> Enum.flat_map(fn c ->
      {cix, cax} = Map.get(c, "x")
      {ciy, cay} = Map.get(c, "y")
      for x <- (cix..cax), y <- (ciy..cay) do
        {{x, y}, "#"}
      end
    end)

    data = Map.merge(Map.new(base_data), Map.new(clay_data))
    spring = {500, min_y - 1}
    data = Map.put(data, spring, "+")
    view = {min_x, max_x, min_y, max_y}
    actual = {min_x - 1, max_x + 1, min_y - 1, max_y}
    %Scan{data: data, view: view, actual: actual, spring: spring}
  end

  def print_scan(scan) do
    {min_x, max_x, min_y, max_y} = scan.actual
    for y <- min_y..max_y do
      line = Enum.join(min_x..max_x |> Enum.map(fn x -> Map.get(scan.data, {x, y}) end), "")
      IO.puts line
    end
    IO.puts ""
  end

  def find_type(start, dir, type_fn, data) do
    {sx, sy} = start
    {fxy, mm} = case dir do
      :up -> {fn {x, y} -> x == sx and y <= sy end, &Enum.max/2}
      :down -> {fn {x, y} -> x == sx and y >= sy end, &Enum.min/2}
      :left -> {fn {x, y} -> y == sy and x <= sx end, &Enum.max/2}
      :right -> {fn {x, y} -> y == sy and x >= sx end, &Enum.min/2}
      _ -> IO.puts("Bad dir " <> Atom.to_string(dir))
    end

    coords = Map.to_list(data) |> Enum.filter(fn {xy, c} -> fxy.(xy) and type_fn.(c) end) |>
      Enum.map(fn {kv, _} -> kv end)
    mm.(coords, fn -> nil end)
  end

  def find_barrier(start, dir, data) do
    type_fn = fn c -> c == "#" || c == "~" end
    find_type(start, dir, type_fn, data)
  end

  def find_empty(start, dir, data) do
    type_fn = fn c -> c == "." || c == "|" end
    find_type(start, dir, type_fn, data)
  end

  # fills from start through fin with c
  def fill_between(start, fin, c, data) do
    {sx, sy} = start
    {ex, ey} = fin
    coords = Map.keys(data) |> Enum.filter(fn {x, y} ->
      if sx <= ex do
        x >= sx and x <= ex
      else
        x >= ex and x <= sx
      end
      and
      if sy <= ey do
        y >= sy and y <= ey
      else
        y >= ey and y <= sy
      end
    end)
    Map.merge(data, coords |> Enum.map(fn kv -> {kv, c} end) |> Map.new)
  end

  def above({x, y}) do
    {x, y-1}
  end

  def below({x, y}) do
    {x, y+1}
  end

  def left({x, y}) do
    {x-1, y}
  end

  def right({x, y}) do
    {x+1, y}
  end

  def go_down(start, data, max_y) do
    if verbose() > 1 do
      IO.inspect ["Going down from ", start]
    end
    {sx, _} = start
    floor = find_barrier(start, :down, data)
    if floor != nil do
      if verbose() > 1 do
        IO.inspect ["Floor found, filling down to ", above(floor)]
      end
      {above(floor), fill_between(start, above(floor), "|", data)}
    else
      if verbose() > 1 do
        IO.inspect ["No floor found, filling down to ", max_y]
      end
      {nil, fill_between(start, {sx, max_y}, "|", data)}
    end
  end

  def is_sideways_of(a, b, dir) do
    {ax, _} = a
    {bx, _} = b
    if dir == :left do
      ax < bx
    else
      ax > bx
    end
  end

  def go_sideways(start, dir, data) do
    wall = find_barrier(start, dir, data)
    above_gap = case find_empty(below(start), dir, data) do
      nil -> nil
      g -> above(g)
    end
    # gap is always non-nil because the left/rightmost x column is always empty
    if wall == nil or is_sideways_of(wall, above_gap, dir) do
      {nil, above_gap, fill_between(start, above_gap, "|", data)}
    else
      fin = if dir == :left, do: right(wall), else: left(wall)
      {wall, nil, fill_between(start, fin, "|", data)}
    end
  end

  def send_drop(start, data, max_y) do
    {drop, data} = go_down(start, data, max_y)
    if drop != nil do
      if verbose() > 1 do
        IO.inspect ["Found drop: ", drop]
      end
      {left_wall, left_gap, data} = go_sideways(drop, :left, data)
      {right_wall, right_gap, data} = go_sideways(drop, :right, data)
      if left_gap == nil and right_gap == nil do
        fill_between(right(left_wall), left(right_wall), "~", data)
      else
        data = if left_gap != nil do
          send_drop(left_gap, data, max_y)
        else
          data
        end
        if right_gap != nil do
          send_drop(right_gap, data, max_y)
        else
          data
        end
      end
    else
      if verbose() > 1 do
        IO.puts "No floor found"
      end
      data
    end
  end

  def send_drops_until_stable(scan) do
    {_, _, _, max_y} = scan.actual
    if verbose() > 0 do
      IO.puts "Sending a drop"
    end
    new_scan = %{scan | data: send_drop(below(scan.spring), scan.data, max_y)}
    if new_scan == scan do
      scan
    else
      if verbose() > 1 do
        print_scan(new_scan)
      end
      send_drops_until_stable(new_scan)
    end
  end

  def count_damp(scan) do
    {min_x, max_x, min_y, max_y} = scan.view
    Map.to_list(scan.data) |>
      Enum.filter(fn {{x, y}, c} -> x >= min_x and x <= max_x and
                                    y >= min_y and y <= max_y and
                                    (c == "~" or c == "|") end) |>
      Enum.count
  end

  def main(file) do
    scan = parse_input(file)
    print_scan(scan)
    final_scan = send_drops_until_stable(scan)
    print_scan(final_scan)
    damp_num = count_damp(final_scan)
    IO.puts "Damp squares: " <> Integer.to_string(damp_num)
  end
end

Solve.main(Enum.at(System.argv(), 0))
